### PR TITLE
chore: scene loading fixes and performance improvements

### PR DESCRIPTION
**Pull request by eordano**
Drop custom logic from a time when multiple scenes were pushed at the same time.

### CODE REVIEW

**Review by matiasbargas — DISMISSED** [review]

**Comment by eordano** [comment]
Despite the fact that I love this change, due to miscommunications with the Unity client unloading scenes when you're too far off, plus lack of communication of offloading events, plus not triggering reloads upon new messages, I found a bug by which there are some scenes that are not reloaded "when you come back". Marked as blocked for the time being

**Comment by eordano** [comment]
Fixed. Made some cleanup of tech debt here and there.

File: packages/entryPoints/editor.ts [hunk 1]
```diff
@@ -78,12 +83,10 @@ async function loadScene(scene: IScene & { baseUrl: string }) {
 }
 
 async function initializePreview(userScene: EnvironmentData<LoadableParcelScene>, parcelCount: number) {
-  loadedParcelSceneWorkers.forEach($ => {
-    $.dispose()
-    loadedParcelSceneWorkers.delete($)
-  })
+  loadedSceneWorkers.forEach($ => forceStopParcelSceneWorker($))
```
Comment (menduz) [inline packages/entryPoints/editor.ts:86]: `forceStopParcelSceneWorker($)` should check `if ($.persistent) return`

File: packages/shared/world/parcelSceneManager.ts [hunk 2]
```diff
@@ -14,41 +15,80 @@ export type EnableParcelSceneLoadingOptions = {
   onLoadParcelScenes?(x: ILand[]): void
 }
 
-export const loadedParcelSceneWorkers: Set<SceneWorker> = new Set()
+export const loadedParcelSceneWorkers: { [base: string]: SceneWorker } = {}
```
Comment (menduz) [inline packages/shared/world/parcelSceneManager.ts]: prefer `Map<string, SceneWorker>` over dictionaries

File: packages/shared/world/parcelSceneManager.ts [hunk 3]
```diff
@@ -14,41 +15,80 @@ export type EnableParcelSceneLoadingOptions = {
...
+
+const sceneWorkerByBaseCoordinate: { [base: string]: SceneWorker } = {}
```
Comment (menduz) [inline packages/shared/world/parcelSceneManager.ts]: same here

File: packages/shared/world/parcelSceneManager.ts [hunk 4]
```diff
@@ -14,41 +15,80 @@ export type EnableParcelSceneLoadingOptions = {
...
+    worker.parcelScene &&
+    worker.parcelScene.data &&
+    worker.parcelScene.data.data &&
+    worker.parcelScene.data.data.scene &&
+    worker.parcelScene.data.data.scene.base
+  )
+}
+
+export function stopParcelSceneWorker(worker: SceneWorker) {
+  if (worker && !worker.persistent) {
```
Comment (menduz) [inline packages/shared/world/parcelSceneManager.ts:54]: the name `stopParcelSceneWorker` does not clearly speak about the condition `worker.persistent` since it is a oneLiner I think we can not use those functions, for clarity.

`forceStopParcelSceneWorker` is a good onw

File: packages/shared/world/parcelSceneManager.ts [hunk 5]
```diff
@@ -14,41 +15,80 @@ export type EnableParcelSceneLoadingOptions = {
...
+  const id = getParcelSceneRootCID(parcelScene)
+  if (loadedParcelSceneWorkers[id]) {
+    return loadedParcelSceneWorkers[id]
+  }
+  const parcelSceneWorker = new SceneWorker(parcelScene, transport)
+
+  if (parcelSceneWorker) {
+    loadedParcelSceneWorkers[getParcelSceneRootCID(parcelScene)] = parcelSceneWorker
+
+    if ('sendUserViewMatrix' in parcelSceneWorker) {
```
Comment (menduz) [inline packages/shared/world/parcelSceneManager.ts:82]: Now that I see it, it should be a worker's responsibility to subscribe/desubscribe to the positionObservable. Can we move that to the constructor/dispose methods to reduce coupling in this module? Up until now I thought it was for loading/unloading life cycle managing

File: packages/shared/world/parcelSceneManager.ts [hunk 6]
```diff
@@ -78,14 +113,9 @@ export async function enableParcelSceneLoading(network: ETHEREUM_NETWORK, option
...
-    const parcelSceneToUnload = await ret.getParcelData(sceneCID)
-    loadedParcelSceneWorkers.forEach($ => {
-      if (!$.persistent && getParcelSceneCID($) === parcelSceneToUnload.mappingsResponse.root_cid) {
-        $.dispose()
-        loadedParcelSceneWorkers.delete($)
-      }
-    })
+  ret.on('Scene.shouldUnload', async (opts: { sceneCID: string }) => {
+    const worker = loadedParcelSceneWorkers[opts.sceneCID]
+    stopParcelSceneWorker(worker)
```
Comment (menduz) [inline packages/shared/world/parcelSceneManager.ts:121]: what if `!worker`?

File: packages/unity-interface/dcl.ts [hunk 7]
```diff
@@ -282,20 +267,19 @@ async function initializeDecentralandUI() {
 
   await ensureUiApis(worker)
 
-  loadedParcelSceneWorkers.add(worker)
+  loadedParcelSceneWorkers['ui'] = worker
```
Comment (menduz) [inline packages/unity-interface/dcl.ts]: wat

**Review by menduz — DISMISSED** [review]

File: packages/decentraland-loader/lifecycle/controllers/parcel.ts [hunk 8]
```diff
@@ -33,6 +33,7 @@ export class ParcelLifeCycleController extends EventEmitter {
     const sightedParcels = parcelsInScope(this.config, position)
     const sightedParcelsDict: { [key: string]: boolean } = {}
     for (const parcel of sightedParcels) {
+      sightedParcelsDict[parcel] = true
```
Comment (fmiras) [inline packages/decentraland-loader/lifecycle/controllers/parcel.ts]: Wouldn't this array be better as a `Set` type?

File: packages/decentraland-loader/lifecycle/controllers/scene.ts [hunk 9]
```diff
@@ -35,14 +35,16 @@ export class SceneLifeCycleController extends EventEmitter {
   async onSight(position: string) {
     let sceneCID = await this.requestSceneCID(position)
     if (sceneCID) {
-      this.sceneParcelSightCount[sceneCID] = this.sceneParcelSightCount[sceneCID] || 1
+      this.sceneParcelSightCount[sceneCID] = this.sceneParcelSightCount[sceneCID] || 0
+      this.sceneParcelSightCount[sceneCID]++
```
Comment (fmiras) [inline packages/decentraland-loader/lifecycle/controllers/scene.ts]: I'm seeing that this object is only for counting. Does the state of the class persists when, for example, teleporting? If not we should clear this array right? I suggest using `Map` instead of plain object/array. It's better if we later need to delete elements.

File: packages/entryPoints/editor.ts [hunk 10]
```diff
@@ -121,9 +124,6 @@ async function initializePreview(userScene: EnvironmentData<LoadableParcelScene>
     evtEmitter.emit('entityBackInScene', e)
   })
 
-  // we need closeParcelScenes to enable interactions in preview mode
-  loadedParcelSceneWorkers.add(parcelScene)
```
Comment (fmiras) [inline packages/entryPoints/editor.ts:125]: Is this entrypoint tested with Builder?

File: packages/shared/world/parcelSceneManager.ts [hunk 11]
```diff
@@ -14,41 +15,80 @@ export type EnableParcelSceneLoadingOptions = {
...
+  worker.dispose()
+  const rootCID = getParcelSceneRootCID(worker.parcelScene)
+  const parcelSceneWorker = loadedParcelSceneWorkers[rootCID]
+  delete loadedParcelSceneWorkers[rootCID]
+  const baseCoordinates = getBaseCoordinates(parcelSceneWorker)
+  if (baseCoordinates) {
+    delete sceneWorkerByBaseCoordinate[baseCoordinates]
+  }
+  userViewMatrixWorkers = userViewMatrixWorkers.filter($ => $ !== worker)
+  loadedSceneWorkers = loadedSceneWorkers.filter($ => $ !== worker)
```
Comment (fmiras) [inline packages/shared/world/parcelSceneManager.ts]: Same here, if we only want to add and remove elements the `Set` or `Map` structures definitely have better performance + we can use `const` instead of `let` in the definition and work with more accurate data

File: packages/shared/world/parcelSceneManager.ts [hunk 12]
```diff
@@ -114,10 +144,8 @@ export function enablePositionReporting() {
 
   positionObservable.add(obj => {
     worldToGrid(obj.position, position)
-    for (let parcelSceneWorker of loadedParcelSceneWorkers) {
-      if (parcelSceneWorker && 'sendUserViewMatrix' in parcelSceneWorker) {
-        parcelSceneWorker.sendUserViewMatrix(obj)
-      }
+    for (let parcelSceneWorker of userViewMatrixWorkers) {
```
Comment (fmiras) [inline packages/shared/world/parcelSceneManager.ts:150]: I still don't like using ES6 `for` statements in some places and lambda functions in other places, can we at least write something in dcl standards and explain the "what" and "why". If you tell me those I can write them down.

CC @menduz

**Review by fmiras — DISMISSED** [review]

File: packages/shared/world/parcelSceneManager.ts (hunk 3 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts]: thanks, will update

File: packages/shared/world/parcelSceneManager.ts (hunk 4 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts:54]: Hmm, isn't that a "leaky" abstraction thing? The implementation being a "one liner" shouldn't change the description. I'll add some docs. This is the function that should be used 99% of cases -- "forceStop" is for exceptional things.

Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts:54]: Rationale:
`stopParcelSceneWorker` stops the worker if it's not persistent
`forceStopParcelSceneWorker` doesn't care if it's perssistent

File: packages/shared/world/parcelSceneManager.ts (hunk 5 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts:82]: TBH i'd like to keep the purpose of the PR scoped to the critical stuff -- I made changes in how do we do this because I feel it was an improvement in perf but I agree with you...

File: packages/shared/world/parcelSceneManager.ts (hunk 6 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts:121]: `stopParcelSceneWorker` handles that case

File: packages/shared/world/parcelSceneManager.ts (hunk 2 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts]: roger that

File: packages/decentraland-loader/lifecycle/controllers/scene.ts (hunk 9 above)
Comment (eordano) [reply packages/decentraland-loader/lifecycle/controllers/scene.ts]: Yes, that is handled by https://github.com/decentraland/explorer/pull/260/files/ca532f4db0b07accaaf0a5f58243d9096216edb7#diff-f986d78a257e5ba8e8887ce2ae54bb6eR44 (you'll get "onSight" and "lostSight" calls. I'll add some tests to clarify.

File: packages/entryPoints/editor.ts (hunk 10 above)
Comment (eordano) [reply packages/entryPoints/editor.ts:125]: I haven't. Will do, thank you for the review

File: packages/shared/world/parcelSceneManager.ts (hunk 11 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts]: Roger that

File: packages/shared/world/parcelSceneManager.ts (hunk 12 above)
Comment (eordano) [reply packages/shared/world/parcelSceneManager.ts:150]: Let me know which you prefer. I like these kind of statements, they are more semantic than using lambdas

File: packages/unity-interface/dcl.ts (hunk 7 above)
Comment (eordano) [reply packages/unity-interface/dcl.ts]: Thanks. should be `loadParcelScene(worker)`

File: packages/shared/world/parcelSceneManager.ts (hunk 12 above)
Comment (menduz) [reply packages/shared/world/parcelSceneManager.ts:150]: @fmiras `for .. of ..` loops are useful for iterators where the "forEach" does not exist. The performance for critical applications is lower tho. you can also leverage `continue` and `break` statements.

`something.forEach(...)` is usually faster, depending on the free variables of the lamda passed it can be better or worse in performance, counting the frames pushed to the stack.

anyways, analyzing it is a deep dive only worthy for performance critical applications.

TLDR: make it readable and understandable unless it gets called thousands of times

File: packages/shared/world/parcelSceneManager.ts [hunk 13]
```diff
@@ -14,41 +15,80 @@ export type EnableParcelSceneLoadingOptions = {
...
+
+export function loadParcelScene(parcelScene: ParcelSceneAPI, transport?: ScriptingTransport) {
+  const id = getParcelSceneRootCID(parcelScene)
+  if (loadedParcelSceneWorkers[id]) {
+    return loadedParcelSceneWorkers[id]
+  }
+  const parcelSceneWorker = new SceneWorker(parcelScene, transport)
+
+  if (parcelSceneWorker) {
+    loadedParcelSceneWorkers[getParcelSceneRootCID(parcelScene)] = parcelSceneWorker
```
Comment (matiasbargas) [inline packages/shared/world/parcelSceneManager.ts]: is this `getParcelSceneRootCID` same as `id` above?

**Review by matiasbargas — APPROVED** [review]
1 small comment but I agree with the general idea. Approve waiting for solving all the comments.

**Comment by eordano** [comment]
I changed it to rootcid to explicitly use rootcid. I have no idea in
general if "ID" means location, scene cid, or root cid. I want to start
using scene cid and this is the first step to clean up for that change

File: packages/entryPoints/editor.ts (hunk 10 above)
Comment (eordano) [reply packages/entryPoints/editor.ts:125]: (I see your concern: this is now done by `loadParcelScene` -- it was a leak on the abstraction)

**Comment by menduz** [comment]
Looks good, is the preview for the CLI tested? this code touches code related to the scene reloading

**Comment by eordano** [comment]
> Looks good, is the preview for the CLI tested? this code touches code related to the scene reloading

Reloading is disabled ATM. The current behavior should be the same: if you're asking about the missing `loadedParcelScenes.add` that is now a responsibility of the `loadParcelScene` function.

PS: Yes, I tested the CLI and it works

**Review by menduz — APPROVED** [review]

File: packages/entryPoints/editor.ts (hunk 10 above)
Comment (eordano) [reply packages/entryPoints/editor.ts:125]: Confirmed it works as expected

**Comment by eordano** [comment]
@fmiras please review and merge

**Review by fmiras — APPROVED** [review]
